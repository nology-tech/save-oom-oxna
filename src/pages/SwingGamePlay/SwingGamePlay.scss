@use '../../styles/variables' as *;
@use '../../../node_modules/animate.css/animate.css';

.swing-game-play {
  margin-left: auto;
  margin-right: auto;
  width: 100%;
  background-image: url(../../assets/images/GameBackground.png);
  background-size: 100% 100%;
  background-repeat: no-repeat;

  &__phonic {
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  &__oom {
    position: absolute;
    width: 14rem;
    height: 14rem;
    bottom: 0.5%;
    left: 65%;
    -webkit-animation-duration: 7s;
    animation-duration: 7s;
  }

  &__squirrel {
    position: absolute;
    width: 6rem;
    height: 6rem;
    cursor: pointer;

    bottom: 5%;
    left: 30%;
    -webkit-animation-duration: 2s;
    animation-duration: 2s;
  }

  &__score {
    position: absolute;
    bottom: 5%;
    left: 45%;
    width: 25rem;
  }
}

.animate__swinging {
  -webkit-transform-origin: top center;
  transform-origin: top center;
  -webkit-animation-name: swinging;
  animation-name: swinging;
}

@-webkit-keyframes swinging {
  0% {
    -webkit-transform: rotate(20deg);
  }
  50% {
    -webkit-transform: rotate(-15deg);
  }
  75% {
    -webkit-transform: rotate(50deg);
  }
  100% {
    -webkit-transform: rotate(-15deg);
  }
}

@keyframes swinging {
  0% {
    transform: rotate(20deg);
  }
  50% {
    transform: rotate(-15deg);
  }
  75% {
    transform: rotate(50deg);
  }
  100% {
    transform: rotate(-15deg);
  }
}

.animate__animated.animate__fastest {
  -webkit-animation-duration: calc(1s * 0.01);
  animation-duration: calc(1s * 0.01);
  -webkit-animation-duration: calc(var(--animate-duration) * 0.01);
  animation-duration: calc(var(--animate-duration) * 0.01);
}

@media screen and (min-width: 1100px) {
  .swing-game-play {
    &__oom {
      width: 18rem;
      height: 18rem;
    }
    &__squirrel {
      width: 10rem;
      height: 10rem;
    }
  }
}
