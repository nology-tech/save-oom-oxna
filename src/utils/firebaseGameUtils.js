import { db } from "../firebase";
import {
  addDoc,
  collection,
  doc,
  getDocs,
  getDoc,
  query,
  where,
  orderBy,
  Timestamp,
} from "firebase/firestore";


/**
 * Gets a user object using the userId.
 * @param {*} userId 
 * @returns 
 */
export const getUserById = async( userId ) => {
  const docRef = doc(db, "users", userId);
  const docSnap = await getDoc( docRef );
  return docSnap;
}

/** 
 * Get all the rounds of a specific game for a given user, ordered by most recent first 
 @param {*} userId
*@returns a querySnap (document snapshot), returns a Promise.
*/

export const getGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** 
   * Get all the correct round of a specific game for a given user, ordered by most recent first 
  @param {String} userId
  @param {String} gameId
  *@returns a promise with the query snapshot.
  */
export const getCorrectGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    where("correct", "==", true),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** Get all the incorrect round of a specific game for a given user, ordered by most recent first 
  @param {String} userId
  @param {String} gameId
  *@returns  a promise with the query snapshot.
  */
export const getIncorrectGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    where("correct", "==", false),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** Function that returns an array of rounds 
  @param {String} userId
  @param {String} gameId
  @param {Function} getGameRounds
  *@returns returns a promise.
  */
export const getArrayOfRounds = async (id, game, getGameRounds) => {
  const querySnap = await getGameRounds(id, game);
  const phonicsArr = querySnap.docs.map((doc) => {
    return doc.data().phonic;
  });
  return phonicsArr;
};

/** Function that returns an array of rounds 
  @param {String} id
  @param {String} game
  @param {Object} roundInfo
  @param {String} answer
  *@returns returs a promise.
  */
export const saveUserRound = async (id, game, level, gameStats, answer) => {
  const { isCorrect, score } = gameStats;
  await addDoc(collection(db, "users", id, "rounds_played"), {
    correct: isCorrect,
    gameId: game,
    phonic: answer,
    level: level,
    roundTime: Timestamp.fromDate(new Date()).toDate(),
    score: score,
  });
};
