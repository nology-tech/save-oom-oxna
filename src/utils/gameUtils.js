import phonicsData from "../data/phonicsData";
import { 
    // getGameRoundsForUser, 
    getCorrectGameRoundsForUser, 
    getIncorrectGameRoundsForUser,
    getArrayOfRounds
     } from "./firebaseGameUtils";

// const staticPhonicsArray = Object.keys(phonicsData.levelOne);

/**
 * Get a list of phonics from the static array, based on the level.
 * @param {*} level 
 * @returns 
 */
const getStaticPhonicsArray = (level) => {
    let levelData = null;
    switch(level) {
        case 1:
            levelData = "levelOne";
            break;
        case 2:
            levelData = "levelTwo";
            break;
        case 3:
            levelData = "levelThree";
            break;
        case 4:
            levelData = "levelFour";
            break;
        case 5:
            levelData = "levelFive";
            break;
    }
    return Object.keys( phonicsData[levelData] );
}

/**
 * Shuffle that array. Nice and simple algorithm.
 * Code courtesy of Geeks for Geeks.
 * @param {*} array 
 * @returns 
 */
const shuffleArray = (array) => {
    for (var i = array.length - 1; i > 0; i--) {
    
        // Generate random number
        var j = Math.floor(Math.random() * (i + 1));
                    
        var temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
        
    return array;
}
    

/**
 * Get the array of sounds for the swing game.
 * Attempts to return a mix of incorrect, correct and new sounds for the given child, based on their
 * past results. Basic algo at the moment!
 * @param {*} id 
 * @param {*} level 
 * @returns 
 */
export const getArrayForSwing = async (id, level) => {

    // 50 wrong = 8
    // 30 new = 4
    // fill up from rest = 3

    const MAX_PER_GAME = 15;
    const MAX_INCORRECT_PER_GAME = 8;
    const MAX_CORRECT_PER_GAME = 4;

    const game = "swing";

    const staticArray = getStaticPhonicsArray( level );
    console.log(`static array for level ${level}`, staticArray);

    const incorrect = await getArrayOfRounds( id, game, getIncorrectGameRoundsForUser);
    const correct = await getArrayOfRounds( id, game, getCorrectGameRoundsForUser);

    console.log( "correct", correct);
    console.log( "incorrect", incorrect);

    // all results
    const result = new Set();

    // add in at most MAX_INCORRECT_PER_GAME of incorrect answers
    for (let phonic of incorrect) {
        result.add(phonic);
        if( result.length >= MAX_INCORRECT_PER_GAME ) {
            console.log("Break!! Reach max incorrect");
            break;
        }
    }
    console.log("After adding incorrect", result);

    let maxCorrect = result.length + MAX_CORRECT_PER_GAME;
    // add in at most MAX_CORRECT_PER_GAME of correct answers
    for (let phonic of correct) {
        result.add(phonic);
        if( result.length >= maxCorrect ) {
            break;
        }
    }
    console.log("After adding correct", result);

    // fill up remaining
    for (let i = 0; i < staticArray.length; i++) {
        result.add( staticArray[i] );
        if( result.size >= MAX_PER_GAME ) {
            break;
        }
    }

    console.log("At end", result);

    const shuffled = shuffleArray( Array.from(result) );

    return shuffled;
}