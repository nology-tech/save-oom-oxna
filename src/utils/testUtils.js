import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import UserContext from "../contexts/UserContext";

export const customRender = (
  ui,
  useRouting = true,
  contextOptions = { useContext: true, value: undefined }
) => {
  // wrap components in routing and context if requested
  const routingUiResult = useRouting && wrapWithRouting(ui);
  const contextUiResult =
    contextOptions.useContext &&
    wrapWithContext(routingUiResult, contextOptions.value);

  // use RTL's render function to return the test component
  return render(contextUiResult);
};

const wrapWithRouting = (ui) => {
  return <Router>{ui}</Router>;
};

const wrapWithContext = (ui, value) => {
  return <UserContext.Provider value={value}>{ui}</UserContext.Provider>;
};
