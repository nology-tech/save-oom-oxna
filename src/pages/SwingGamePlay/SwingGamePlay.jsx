import React, { useState, useContext, useEffect } from "react";
import swingingOom from "../../assets/images/Group 146swingingOom.png";
import squirrel from "../../assets/images/squirrel.png";
import AnimatedImage from "../../components/AnimatedImage/AnimatedImage";
import ValidateAnswerButtons from "../../components/ValidateAnswerButtons/ValidateAnswerButtons";
import PhonicComponent from "../../components/PhonicComponent/PhonicComponent";
import phonicsData from "../../data/phonicsData";
import GameEnd from "../GameEnd/GameEnd";
import "./SwingGamePlay.scss";
import Timer from "../../components/Timer/Timer";
import shortid from "shortid";
import OomsNeedsContainer from "../../containers/OomsNeedsContainer/OomsNeedsContainer";
import {saveUserRound} from "../../utils/firebaseGameUtils";
import UserContext from "../../contexts/UserContext";
import {getArrayForSwing} from "../../utils/gameUtils";

let gameScore = 0;
const SwingGamePlay = () => {

  const game = "swing";
  // should become props?
  const level = 1;

  let currentUserName = "";
  let currentUserId = 0;
  try {
    const { user } = useContext(UserContext);
    currentUserName = user.name;
    currentUserId = user.userId;

  } catch (e) {
    console.error("Error attempting to get userContext!", e);
  }

  const [gameState, setGameState] = useState({
    score: 0,
    index: 0,
    isCorrect: false,
    isGameOver: false,
    isGameReady: false,
    counter: 0,
  });

  const [hintAnimation, setHintAnimation] = useState(false);
  const [phonicsArray, setPhonicsArray] = useState([]);

  // load up the Phonics array 
  useEffect(() => {
    getArrayForSwing(currentUserId, 1)
      .then( array => {
          let newGameState = { ...gameState };
          newGameState.isGameReady = true;
          setGameState(newGameState);
          setPhonicsArray(array);
      
    })
    .catch((e) => console.error("Error attempting to load phonics array", e));

  }, []  );

  const handleCorrect = () => {
    const currentPhonic = phonicsArray[gameState.index];
    let newGameState = { ...gameState };
    newGameState.isCorrect = true;
    setHintAnimation(false);
    newGameState.score = newGameState.score + 1;
    gameScore = newGameState.score;
    newGameState.counter = newGameState.counter + 1;
    newGameState.index = handleIndexChange();
    setGameState(newGameState);
    console.log(newGameState, gameState, "handleCorrect");
    // save game results
    saveUserRound( currentUserId, game, level, newGameState, currentPhonic)
      .then(() => {console.log("Saved user round - correct!")})
      .catch(() => {console.error("Error attempting to save a user's round")});
  };

  const handleIncorrect = () => {
    const currentPhonic = phonicsArray[gameState.index];
    let newGameState = { ...gameState };
    newGameState.isCorrect = false;
    setHintAnimation(false);
    console.log(newGameState.isCorrect);
    newGameState.index = newGameState.index + 1;
    setGameState(newGameState);
    console.log(newGameState, gameState, "handleIncorrect");
    // save game results
    saveUserRound( currentUserId, game, level, newGameState, currentPhonic)
      .then(() => {console.log("Saved user round - incorrect!")})
      .catch(() => {console.error(`Error attempting to save a user ${currentUserId} round`)});
  };

  const handleHint = () => {
    let newGameState = { ...gameState };
    setHintAnimation(true);
    newGameState.isCorrect = null;
    console.log(newGameState.isCorrect);
    newGameState.counter = newGameState.counter + 1;
    let audio = new Audio(
      phonicsData.levelOne[phonicsArray[gameState.index]].soundUrl
    );
    audio.play();
    setGameState(newGameState);
  };

  const handleIndexChange = () => {
    if (gameState.index < phonicsArray.length) {
      return gameState.index + 1;
    } else {
      return 0;
    }
  };

  const handleGameEnd = () => {
    let newGameState = { ...gameState };
    newGameState.isGameOver = true;
    setGameState(newGameState);
    console.log(newGameState, gameState, 'handleGameEnd');
  };

  const getId = () => {
    const id = shortid.generate();
    console.log(id);
    return id;
  };

  const squirrelAnimationType2 = hintAnimation ? 'animate__bounce' : '';
  const oomAnimationType = gameState.isCorrect ? 'animate__swinging' : '';

  const gameNotAvailable = ! gameState.isGameReady || gameState.isGameOver;
  const gameNotAvailableJsx = gameState.isGameOver ? 
      <GameEnd score={gameScore} childName={currentUserName} /> :
          gameState.isGameReady == false ?
            <p>Nothing!</p> : null;

  return (
    <div className="swing-game-play">
      {gameNotAvailable ? (
        gameNotAvailableJsx
      ) : (
        <>
          <OomsNeedsContainer />
          <div className='swing-game-play__phonic'>
            <Timer startTime={60} handleGameEnd={handleGameEnd} />
            <PhonicComponent phonicText={phonicsArray[gameState.index]} />
          </div>

          <div onClick={handleHint}>
            <AnimatedImage
              key={getId()}
              imageToAnimate={squirrel}
              animationClass={'animate__animated.animate__fastest'}
              animationType={` ${squirrelAnimationType2}`}
              imageStylesClass={'swing-game-play__squirrel'}
            />
          </div>
          <AnimatedImage
            key={getId()}
            imageToAnimate={swingingOom}
            animationClass={'animate__animated.animate__fastest'}
            animationType={oomAnimationType}
            imageStylesClass={'swing-game-play__oom'}
          />

          <ValidateAnswerButtons
            handleCorrect={handleCorrect}
            handleIncorrect={handleIncorrect}
          />

          <p className='swing-game-play__score'>
            Number Of Correct Sounds: {gameState.score}
          </p>
        </>
      )}
    </div>
  );
};

export default SwingGamePlay;
