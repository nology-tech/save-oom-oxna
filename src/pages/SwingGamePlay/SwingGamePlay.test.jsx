import { customRender } from '../../utils/testUtils';
import SwingGamePlay from './SwingGamePlay';
import { render, screen } from '@testing-library/react';

const userFnctn = () => {
  console.log("Updates user context ")
}
const userObj = {
  user: {
    name: "Jack",
    email:"Hello@gmail"
  },
  setUser: userFnctn,
};


it("Should render the Swing Game Play screen", () => {
  const { container } = customRender(<SwingGamePlay />);

  expect(container).toMatchSnapshot();
});

it('Should render the score count screen', () => {
  customRender(<SwingGamePlay />, true, {useContext: true, value: userObj});
  const scoreCount = screen.getByText(/Nothing!/i);
  expect(scoreCount).toBeInTheDocument();
});
