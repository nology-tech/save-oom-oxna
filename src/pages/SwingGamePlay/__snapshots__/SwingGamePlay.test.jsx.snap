// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the Swing Game Play screen 1`] = `
<div>
  <div
    class="swing-game-play"
  >
    <p>
      Nothing!
    </p>
  </div>
</div>
`;
